cupy/_core/include/cupy/complex/arithmetic.h

```diff
@@ -113,6 +113,12 @@ __host__ __device__ inline complex<T> operator/(const complex<T>& lhs,
   return complex<T>(lhs.real() / rhs, lhs.imag() / rhs);
 }
 
+template <typename T, typename S>
+__host__ __device__ inline complex<T> operator/(const complex<T>& lhs,
+                                                const S& rhs) {
+  return complex<T>(lhs.real() / rhs, lhs.imag() / rhs);
+}
+
```

Comment:
@takagi, I saw there were many test failures when we relaxed the dtype of the second arg of `operator/`. So I think it will be good to add another func supporting different dtype args in `operator/`. 

The test cases are passing now. Please see:
![Screenshot from 2022-07-23 18-52-34](https://user-images.githubusercontent.com/62256509/180607046-7e50faa9-ce05-49f8-8fbe-a4d97f349b37.png)

What do you think? Thanks!

Reply:
The errors say we should also change the declaration of operator/ in complex/complex.h(339). But, hum, now I changed my mind that  let's keep the original declaration and definition `template <typename T>
__host__ __device__ inline complex<T> operator/(const complex<T>& lhs, const T& rhs)` and cast the divisor in callers,  `sum / satic_cast<T>(_count)`, as the complex-related files are borrowed from CUDA toolkit and let's keep them as they're.

Reply:
Hi, @takagi! Thanks for your input! I've something in my mind ...

I might be not right. But I think we cannot use `static_cast<T>(_count)` here. Because we haven't used the templated kernel, rather we have explicitly defined that for `complex64` and `complex128` (Thanks to your offline help!!!). `T` is not defined here. And if we cast it to `complex` as we'll get:
```python
error: argument list for class template "thrust::complex" is missing
```
We also can't cast it to use `complex64` and `complex128` as these identifiers are not defined there.

Sounds reasonable to you? Thanks for looking into it :)

Reply:
ah, what in my mind was `_nanvar_core`, and we do not have to change it as it only takes real numbers. You may use `static_cast<float>` for `_nanvar_core_complex64`, and  `static_cast<double>` for `_nanvar_core_complex128`.

Reply:
I think casting the denominator to `float` and `double`, we'll get a compilation error ... (dividing division operator with diff operand types which looks reasonable to me). As you told, **xref**: https://github.com/cupy/cupy/pull/6855#discussion_r921753216
```python
File "/home/khushi/Documents/repo/cupy/cupy/cuda/compiler.py", line 305, in _compile
    compiled_obj, mapping = prog.compile(options, log_stream)
  File "/home/khushi/Documents/repo/cupy/cupy/cuda/compiler.py", line 704, in compile
    raise CompileException(log, self.src, self.name, options,
cupy.cuda.compiler.CompileException: /tmp/tmpatn3spsy/90a190cf40314048aa0bc41e71bb6f79257cd993.cubin.cu(48): error: no operator "/" matches these operands
            operand types are: const T / const long long

1 error detected in the compilation of "/tmp/tmpatn3spsy/90a190cf40314048aa0bc41e71bb6f79257cd993.cubin.cu".
```
I re-checked it. We got the above error. Same as we got earlier.

Reply:
Can I see the lines you changed?

Reply:
Sure, here's it:
```python
-    'nanvar_impl(x, sum / _count, max(_count - ddof, 0LL))',
+    'nanvar_impl(x, sum / static_cast<float>(_count), max(_count - ddof, 0LL))',
 ```
```python
-    'nanvar_impl(x, sum / _count, max(_count - ddof, 0LL))',
+    'nanvar_impl(x, sum / static_cast<double>(_count), max(_count - ddof, 0LL))',
```
Thanks!

Reply:
hmm, also can I have the dumped CUDA C code when you got the error in https://github.com/cupy/cupy/pull/6869#discussion_r928486597?

Reply:
Sure, please see:
<details>
<summary>CUDA source (Click Please)</summary>

<p>

```cuda
01 #include <cupy/complex.cuh>
02 #include <cupy/carray.cuh>
03 #include <cupy/atomics.cuh>
04 
05 typedef complex<float> S;
06 typedef complex<float> T;
07 typedef double U;
08 
09 
10 template <typename S, typename T>
11 __device__ T nanvar_impl(S x, T mean, long long alpha) {
12     return (isnan(x) ? T(0) : T((x - mean) * (x - mean))) / alpha;
13 }
14 
15 template <typename S, typename T>
16 __device__ T nanvar_impl(complex<S> x, complex<T> mean, long long alpha) {
17     return (isnan(x) ? T(0) : T(norm(x - mean))) / alpha;
18 }
19 
20 #define REDUCE(a, b) (a + b)
21 #define POST_MAP(a) (out = a)
22 #define _REDUCE(_offset) if (_tid < _offset) {   _type_reduce _a = _sdata[_tid], _b = _sdata[(_tid + _offset)];   _sdata[_tid] = REDUCE(_a, _b); }
23 
24 typedef U _type_reduce;
25 extern "C" __global__ void _nanvar_core(const CArray<complex<float>, 2, 1, 1> _raw_x, const CArray<complex<float>, 2, 0, 1> _raw_sum, const CArray<long long, 2, 0, 1> _raw__count, const long long ddof, CArray<double, 1, 1, 1> _raw_out, CIndexer<2, 1> _in_ind, CIndexer<1, 1> _out_ind, const int _block_stride) {
26   __shared__ char _sdata_raw[512 * sizeof(_type_reduce)];
27   _type_reduce *_sdata = reinterpret_cast<_type_reduce*>(_sdata_raw);
28   unsigned int _tid = threadIdx.x;
29 
30   int _J_offset = _tid >> __popc(_block_stride - 1);  // _tid / _block_stride
31   ptrdiff_t _j_offset = (ptrdiff_t)_J_offset * _out_ind.size();
32   int _J_stride = 512 >> __popc(_block_stride - 1);
33   ptrdiff_t _j_stride = (ptrdiff_t)_J_stride * _out_ind.size();
34 
35   for (ptrdiff_t _i_base = (ptrdiff_t)blockIdx.x * _block_stride;
36        _i_base < _out_ind.size();
37        _i_base += (ptrdiff_t)gridDim.x * _block_stride) {
38     _type_reduce _s = _type_reduce(0);
39     ptrdiff_t _i =
40         _i_base + (_tid & (_block_stride - 1));  // _tid % _block_stride
41     int _J = _J_offset;
42     for (ptrdiff_t _j = _i + _j_offset; _j < _in_ind.size();
43          _j += _j_stride, _J += _J_stride) {
44       _in_ind.set(_j);
45       const S x = _raw_x[_in_ind.get()];
46 const T sum = _raw_sum[_in_ind.get()];
47 const long long _count = _raw__count[_in_ind.get()];
48       _type_reduce _a = static_cast<_type_reduce>(nanvar_impl(x, sum / _count, max(_count - ddof, 0LL)));
49       _s = REDUCE(_s, _a);
50     }
51     _sdata[_tid] = _s;
52     __syncthreads();
53     for (unsigned int _block = 512 / 2;
54          _block >= _block_stride; _block >>= 1) {
55       if (_tid < _block) {
56         _REDUCE(_block);
57       }
58       __syncthreads();
59     }
60     if (_tid < _block_stride) {
61       _s = _sdata[_tid];
62     }
63     if (_tid < _block_stride && _i < _out_ind.size()) {
64       _out_ind.set(static_cast<ptrdiff_t>(_i));
65       U &out = _raw_out[_out_ind.get()];
66       POST_MAP(_s);
67     }
68   }
69 }
```

</p>
</details>

Reply:
Thanks! Seems that the error is in `_nanvar_core`. Do we still have a path that `_nanvar_core` takes complex numbers?

Reply:
I think there were failures in `nanvar_core_out`. Casting that using `static_cast<T>()` works well. All the test case passes now. I've modified in the latest commit [2e92ef1](https://github.com/cupy/cupy/pull/6869/commits/2e92ef1b30c60aa64deecad46e1a602f27242f35).

Thanks for your help!